src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSProxyContext.cs

```diff
@@ -0,0 +1,448 @@
+        public nint TargetTID;
+        public int ThreadId;
+        public bool IsMainThread;
+        public JSSynchronizationContext SynchronizationContext;
+
+        [UnsafeAccessor(UnsafeAccessorKind.Field, Name = "thread_id")]
+        private static extern ref long GetThreadNativeThreadId(Thread @this);
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static IntPtr GetNativeThreadId()
+        {
+            return (int)GetThreadNativeThreadId(Thread.CurrentThread);
+        }
+
+        // Context of the main thread
+        private static JSProxyContext? _MainInstance;
+        public static JSProxyContext MainInstance
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => _MainInstance!;
+            set => _MainInstance = value;
+        }
+
+        // Context captured from parameters of the current operation.
+        // Most of the time, this matches the current thread.
+        // When we pass `JSObject` instance to `JSImport`, it could have thread affinity different than the current thread.
+        // We will use that parameter to capture the target thread to call into.
+        [ThreadStatic]
+        public static JSProxyContext? CapturedInstance;
+
+        // Context of the current thread. Could be null on threads which don't have JS interop, like managed thread pool threads.
+        [ThreadStatic]
+        public static JSProxyContext? CurrentInstance;
+
+        // This is context to dispatch into. In order of preference
+        // - captured context by arguments of current/pending JSImport call
+        // - current thread context, for calls from JSWebWorker threads with the interop installed
+        // - main thread, for calls from any other thread, like managed thread pool or `new Thread`
+        public static JSProxyContext DefaultInstance
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get
+            {
+                // until we set CapturedInstance in the generated code of JSImport/JSExport, we need to fallback to CurrentInstance to find target thread
+                // because maybe it was not captured by the other parameters yet
+                // after capture is solid, we could drop DefaultInstance and use CapturedInstance instead
+                // TODO: sort generated ToJS() calls to make the capture first
+                return CapturedInstance ?? CurrentInstance ?? MainInstance;
+            }
+        }
+
+        public JSProxyContext(bool isMainThread, JSSynchronizationContext synchronizationContext)
+        {
+            SynchronizationContext = synchronizationContext;
+            Interop.Runtime.InstallWebWorkerInterop();
+            TargetTID = GetNativeThreadId();
+            ThreadId = Thread.CurrentThread.ManagedThreadId;
+            IsMainThread = isMainThread;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool IsTargetThread()
```

Comment:
`IsCurrentThread` is probably a better name for this, and I would probably make it a property. Minor nit though